File: mobile/android/app/src/test/java/com/volttracker/obdpoc/ArchitectureBoundaryTest.java
package com.volttracker.obdpoc;

import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.junit.Test;

public class ArchitectureBoundaryTest {

    private static final List<String> FORBIDDEN_DATA_LAYER_IMPORTS =
            Arrays.asList(
                    "import com.volttracker.obdpoc.MainActivity",
                    "import com.volttracker.obdpoc.VoltBridge",
                    "import com.volttracker.obdpoc.WebViewBootstrap",
                    "import com.volttracker.obdpoc.ObdService",
                    "import com.volttracker.obdpoc.ObdPollingEngine",
                    "import com.volttracker.obdpoc.ElmConnection",
                    "import com.volttracker.obdpoc.DiagnosticScanRunner",
                    "import com.volttracker.obdpoc.ClearDtcRunner",
                    "import com.volttracker.obdpoc.ObdNotifications",
                    "import com.volttracker.obdpoc.PermissionGate",
                    "import com.volttracker.obdpoc.BackupController",
                    "import com.volttracker.obdpoc.DataBackup",
                    "import android.webkit.");

    private static final List<String> FORBIDDEN_ENGINE_UI_REFERENCES =
            Arrays.asList(
                    "import com.volttracker.obdpoc.VoltBridge",
                    "import com.volttracker.obdpoc.WebViewBootstrap",
                    "import android.webkit.",
                    "evaluateJavascript",
                    "webView");

    private static final List<String> FORBIDDEN_SERVICE_WEBVIEW_REFERENCES =
            Arrays.asList(
                    "import android.webkit.",
                    "import com.volttracker.obdpoc.VoltBridge",
                    "import com.volttracker.obdpoc.WebViewBootstrap",
                    "evaluateJavascript");

    @Test
    public void dataLayerDoesNotImportUiServiceOrEngineClasses() throws IOException {
        List<String> violations = new ArrayList<>();
        for (Path sourceRoot : sourceRoots()) {
            Path dataDir =
                    sourceRoot
                            .resolve("com")
                            .resolve("volttracker")
                            .resolve("obdpoc")
                            .resolve("data");
            if (!Files.isDirectory(dataDir)) {
                continue;
            }
            try (java.util.stream.Stream<Path> paths = Files.walk(dataDir)) {
                paths.filter(ArchitectureBoundaryTest::isSourceFile)
                        .forEach(
                                path -> {
                                    try {
                                        String source =
                                                new String(
                                                        Files.readAllBytes(path),
                                                        StandardCharsets.UTF_8);
                                        for (String forbidden : FORBIDDEN_DATA_LAYER_IMPORTS) {
                                            if (source.contains(forbidden)) {
                                                violations.add(
                                                        displayPath(path)
                                                                + " imports upward via "
                                                                + forbidden);
                                            }
                                        }
                                    } catch (IOException ex) {
                                        throw new AssertionError("Could not read " + path, ex);
                                    }
                                });
            }
        }
        assertTrue("Layering violations:\n" + String.join("\n", violations), violations.isEmpty());
    }

    @Test
    public void engineDoesNotReferenceDashboardBridgeOrWebViewApis() throws IOException {
        List<String> violations =
                scanFilesForForbiddenReferences(
                        Arrays.asList(
                                "ObdPollingEngine.java",
                                "ElmConnection.java",
                                "DiagnosticScanRunner.java",
                                "ClearDtcRunner.java"),
                        FORBIDDEN_ENGINE_UI_REFERENCES);

        assertTrue(
                "Engine/UI boundary violations:\n" + String.join("\n", violations),
                violations.isEmpty());
    }

    @Test
    public void serviceLayerDoesNotCallWebViewApis() throws IOException {
        List<String> violations =
                scanFilesForForbiddenReferences(
                        Arrays.asList(
                                "ObdService.java", "ObdNotifications.java", "PermissionGate.java"),
                        FORBIDDEN_SERVICE_WEBVIEW_REFERENCES);

        assertTrue(
                "Service/WebView boundary violations:\n" + String.join("\n", violations),
                violations.isEmpty());
    }

    @Test
    public void dashboardBridgeDoesNotOpenWritableDatabaseDirectly() throws IOException {
        List<String> violations =
                scanFilesForForbiddenReferences(
                        Arrays.asList(
                                "MainActivity.java", "VoltBridge.java", "WebViewBootstrap.java"),
                        Arrays.asList("getWritableDatabase("));

        assertTrue(
                "UI/data boundary violations:\n" + String.join("\n", violations),
                violations.isEmpty());
    }

    private static List<String> scanFilesForForbiddenReferences(
            List<String> fileNames, List<String> forbiddenReferences) throws IOException {
        List<String> violations = new ArrayList<>();

        for (String fileName : fileNames) {
            Path path = sourcePath(fileName);
            String source = new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
            for (String forbidden : forbiddenReferences) {
                if (source.contains(forbidden)) {
                    violations.add(displayPath(path) + " references " + forbidden);
                }
            }
        }
        return violations;
    }

    private static Path sourcePath(String javaFileName) {
        List<String> candidates = new ArrayList<>();
        candidates.add(javaFileName);
        if (javaFileName.endsWith(".java")) {
            candidates.add(
                    javaFileName.substring(0, javaFileName.length() - ".java".length()) + ".kt");
        }

        for (Path sourceRoot : sourceRoots()) {
            Path packageRoot = sourceRoot.resolve("com").resolve("volttracker").resolve("obdpoc");
            for (String candidate : candidates) {
                Path path = packageRoot.resolve(candidate);
                if (Files.isRegularFile(path)) {
                    return path;
                }
            }
        }
        throw new AssertionError("Could not locate Android source file " + javaFileName);
    }

    private static List<Path> sourceRoots() {
        List<Path> roots = new ArrayList<>();
        for (Path root :
                Arrays.asList(
                        Paths.get("src/main/java"),
                        Paths.get("src/main/kotlin"),
                        Paths.get("app/src/main/java"),
                        Paths.get("app/src/main/kotlin"))) {
            if (Files.isDirectory(root)) {
                roots.add(root);
            }
        }
        if (roots.isEmpty()) {
            throw new AssertionError(
                    "Could not locate Android source root from " + Paths.get("").toAbsolutePath());
        }
        return roots;
    }

    private static boolean isSourceFile(Path path) {
        String name = path.toString();
        return name.endsWith(".java") || name.endsWith(".kt");
    }

    private static String displayPath(Path path) {
        for (Path sourceRoot : sourceRoots()) {
            if (path.startsWith(sourceRoot)) {
                return sourceRoot.relativize(path).toString();
            }
        }
        return path.toString();
    }
}

File: mobile/android/app/src/test/java/com/volttracker/obdpoc/ArchitectureBoundaryTest.kt
package com.volttracker.obdpoc

import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

class ArchitectureBoundaryTest {
    @Test
    fun dataLayerDoesNotImportUiServiceOrEngineClasses() {
        val violations =
            scanPackageDirForForbiddenReferences("data", FORBIDDEN_DATA_LAYER_IMPORTS)

        assertTrue("Layering violations:\n" + violations.joinToString("\n"), violations.isEmpty())
    }

    @Test
    fun engineDoesNotReferenceDashboardBridgeOrWebViewApis() {
        val violations =
            scanPackageDirForForbiddenReferences("engine", FORBIDDEN_ENGINE_UI_REFERENCES)

        assertTrue(
            "Engine/UI boundary violations:\n" + violations.joinToString("\n"),
            violations.isEmpty(),
        )
    }

    @Test
    fun serviceLayerDoesNotCallWebViewApis() {
        val violations =
            scanPackageDirForForbiddenReferences("service", FORBIDDEN_SERVICE_WEBVIEW_REFERENCES)

        assertTrue(
            "Service/WebView boundary violations:\n" + violations.joinToString("\n"),
            violations.isEmpty(),
        )
    }

    /**
     * The engine/service rules above are scoped structurally to their package
     * directories. If a refactor accidentally moved those files back into the
     * root package the rules would pass vacuously, so pin the directories as
     * non-empty here.
     */
    @Test
    fun engineAndServicePackagesAreNonEmpty() {
        for (packageDirName in listOf("engine", "service")) {
            assertTrue(
                "Expected at least one source file under the '$packageDirName' package; " +
                    "the boundary rules for it would be vacuous otherwise",
                countPackageSourceFiles(packageDirName) > 0,
            )
        }
    }

    @Test
    fun productionCodeUsesTheSharedLogTagWithoutDependingOnMainActivity() {
        val violations =
            scanProductionFilesForForbiddenReferences(
                { true },
                listOf("MainActivity.TAG"),
            )

        assertTrue(
            "Shared logging must not couple background code to MainActivity:\n" +
                violations.joinToString("\n"),
            violations.isEmpty(),
        )
    }

    @Test
    fun dashboardBridgeDoesNotOpenWritableDatabaseDirectly() {
        val violations =
            scanFilesForForbiddenReferences(
                listOf("MainActivity.kt", "VoltBridge.kt", "WebViewBootstrap.kt"),
                listOf("getWritableDatabase("),
            )

        assertTrue(
            "UI/data boundary violations:\n" + violations.joinToString("\n"),
            violations.isEmpty(),
        )
    }

    @Test
    fun widgetLayerDoesNotImportWebViewOrBridgeApis() {
        val violations = ArrayList<String>()
        for (sourceRoot in sourceRoots()) {
            val widgetDir =
                sourceRoot
                    .resolve("com")
                    .resolve("volttracker")
                    .resolve("obdpoc")
                    .resolve("widget")
            if (!Files.isDirectory(widgetDir)) {
                continue
            }
            Files.walk(widgetDir).use { paths ->
                paths
                    .filter { isSourceFile(it) }
                    .forEach { path ->
                        val source = String(Files.readAllBytes(path), StandardCharsets.UTF_8)
                        for (forbidden in FORBIDDEN_WIDGET_REFERENCES) {
                            if (source.contains(forbidden)) {
                                violations.add(displayPath(path) + " references " + forbidden)
                            }
                        }
                    }
            }
        }
        assertTrue(
            "Widget must stay a thin UI surface (no WebView/bridge/engine deps):\n" +
                violations.joinToString("\n"),
            violations.isEmpty(),
        )
    }

    @Test
    fun dashboardBridgeDependsOnHostSeamNotConcreteActivity() {
        val violations =
            scanFilesForForbiddenReferences(
                listOf("VoltBridge.kt"),
                FORBIDDEN_BRIDGE_ACTIVITY_REFERENCES,
            )

        assertTrue(
            "VoltBridge must depend on the DashboardHost seam, not the concrete MainActivity:\n" +
                violations.joinToString("\n"),
            violations.isEmpty(),
        )
    }

    private companion object {
        val FORBIDDEN_BRIDGE_ACTIVITY_REFERENCES =
            listOf(
                "MainActivity",
            )

        val FORBIDDEN_WIDGET_REFERENCES =
            listOf(
                "import android.webkit.",
                "import com.volttracker.obdpoc.VoltBridge",
                "import com.volttracker.obdpoc.WebViewBootstrap",
                "import com.volttracker.obdpoc.engine.",
                "evaluateJavascript",
            )

        val FORBIDDEN_DATA_LAYER_IMPORTS =
            listOf(
                "import com.volttracker.obdpoc.MainActivity",
                "import com.volttracker.obdpoc.VoltBridge",
                "import com.volttracker.obdpoc.WebViewBootstrap",
                "import com.volttracker.obdpoc.engine.",
                "import com.volttracker.obdpoc.service.",
                "import com.volttracker.obdpoc.BackupController",
                "import com.volttracker.obdpoc.DataBackup",
                "import android.webkit.",
            )

        val FORBIDDEN_ENGINE_UI_REFERENCES =
            listOf(
                "import com.volttracker.obdpoc.VoltBridge",
                "import com.volttracker.obdpoc.WebViewBootstrap",
                "import android.webkit.",
                "evaluateJavascript",
                "webView",
            )

        val FORBIDDEN_SERVICE_WEBVIEW_REFERENCES =
            listOf(
                "import android.webkit.",
                "import com.volttracker.obdpoc.VoltBridge",
                "import com.volttracker.obdpoc.WebViewBootstrap",
                "evaluateJavascript",
            )

        @Throws(IOException::class)
        fun scanFilesForForbiddenReferences(
            fileNames: List<String>,
            forbiddenReferences: List<String>,
        ): List<String> {
            val violations = ArrayList<String>()

            for (fileName in fileNames) {
                val path = sourcePath(fileName)
                val source = String(Files.readAllBytes(path), StandardCharsets.UTF_8)
                for (forbidden in forbiddenReferences) {
                    if (source.contains(forbidden)) {
                        violations.add(displayPath(path) + " references " + forbidden)
                    }
                }
            }
            return violations
        }

        /**
         * Walks every source file under the given subpackage directory of
         * com.volttracker.obdpoc (across all production source roots) and reports
         * files containing any of the forbidden references. Structural scoping:
         * membership in the rule set is decided by package directory, never by
         * filename convention.
         */
        @Throws(IOException::class)
        fun scanPackageDirForForbiddenReferences(
            packageDirName: String,
            forbiddenReferences: List<String>,
        ): List<String> {
            val violations = ArrayList<String>()
            forEachPackageSourceFile(packageDirName) { path ->
                val source =
                    try {
                        String(Files.readAllBytes(path), StandardCharsets.UTF_8)
                    } catch (ex: IOException) {
                        throw AssertionError("Could not read $path", ex)
                    }
                for (forbidden in forbiddenReferences) {
                    if (source.contains(forbidden)) {
                        violations.add(displayPath(path) + " references " + forbidden)
                    }
                }
            }
            return violations
        }

        fun countPackageSourceFiles(packageDirName: String): Int {
            var count = 0
            forEachPackageSourceFile(packageDirName) { count++ }
            return count
        }

        fun forEachPackageSourceFile(
            packageDirName: String,
            action: (Path) -> Unit,
        ) {
            for (sourceRoot in sourceRoots()) {
                val packageDir =
                    sourceRoot
                        .resolve("com")
                        .resolve("volttracker")
                        .resolve("obdpoc")
                        .resolve(packageDirName)
                if (!Files.isDirectory(packageDir)) {
                    continue
                }
                Files.walk(packageDir).use { paths ->
                    paths.filter { isSourceFile(it) }.forEach { action(it) }
                }
            }
        }

        @Throws(IOException::class)
        fun scanProductionFilesForForbiddenReferences(
            includeName: (String) -> Boolean,
            forbiddenReferences: List<String>,
        ): List<String> {
            val violations = ArrayList<String>()
            for (sourceRoot in sourceRoots()) {
                val packageRoot =
                    sourceRoot.resolve("com").resolve("volttracker").resolve("obdpoc")
                if (!Files.isDirectory(packageRoot)) continue
                Files.walk(packageRoot).use { paths ->
                    paths
                        .filter { isSourceFile(it) && includeName(it.fileName.toString()) }
                        .forEach { path ->
                            val source = String(Files.readAllBytes(path), StandardCharsets.UTF_8)
                            for (forbidden in forbiddenReferences) {
                                if (source.contains(forbidden)) {
                                    violations.add(displayPath(path) + " references " + forbidden)
                                }
                            }
                        }
                }
            }
            return violations
        }

        fun sourcePath(sourceFileName: String): Path {
            val candidates = ArrayList<String>()
            candidates.add(sourceFileName)
            if (sourceFileName.endsWith(".java")) {
                candidates.add(
                    sourceFileName.substring(0, sourceFileName.length - ".java".length) + ".kt",
                )
            }

            for (sourceRoot in sourceRoots()) {
                val packageRoot =
                    sourceRoot.resolve("com").resolve("volttracker").resolve("obdpoc")
                for (candidate in candidates) {
                    val path = packageRoot.resolve(candidate)
                    if (Files.isRegularFile(path)) {
                        return path
                    }
                }
            }
            throw AssertionError("Could not locate Android source file $sourceFileName")
        }

        fun sourceRoots(): List<Path> {
            val roots = ArrayList<Path>()
            for (
            root in
            listOf(
                Paths.get("src/main/java"),
                Paths.get("src/main/kotlin"),
                Paths.get("app/src/main/java"),
                Paths.get("app/src/main/kotlin"),
            )
            ) {
                if (Files.isDirectory(root)) {
                    roots.add(root)
                }
            }
            if (roots.isEmpty()) {
                throw AssertionError(
                    "Could not locate Android source root from " +
                        Paths.get("").toAbsolutePath(),
                )
            }
            return roots
        }

        fun isSourceFile(path: Path): Boolean {
            val name = path.toString()
            return name.endsWith(".java") || name.endsWith(".kt")
        }

        fun displayPath(path: Path): String {
            for (sourceRoot in sourceRoots()) {
                if (path.startsWith(sourceRoot)) {
                    return sourceRoot.relativize(path).toString()
                }
            }
            return path.toString()
        }
    }
}
